### PR TITLE
Azure Datalake Sink

### PR DESCRIPTION
**Pull request by davidsloan**
## Connector Overview

The Kafka sink connector for Azure Data Lake is a powerful tool that allows you to seamlessly stream data from Apache Kafka topics into your Azure Data Lake storage. This connector facilitates real-time data integration and analytics, making it a crucial component of data pipelines. It supports various data formats, compression options, and error handling policies to fit your specific use case.

## Key Features

- **Real-time Data Streaming**: The connector enables real-time data streaming from Kafka topics to Azure Data Lake, ensuring that your data is readily available for analysis and processing.

- **Data Format Support**: You can choose the data format that best suits your needs, including Avro and Parquet, and configure compression settings.

- **Error Handling**: The connector offers configurable error handling policies, allowing you to determine how errors are managed during data transfer. You can choose to ignore errors (NOOP), propagate them (THROW), or retry (RETRY) with defined retry limits.

- **Flexible Configuration**: The connector provides various configuration options to adapt to your specific Azure Data Lake setup and requirements. This includes authentication modes, connection settings, and more.

- **File Handling**: You can configure local temporary directories, padding, and flush behavior to optimize the storage process.

## Connector Configuration

Here's an overview of the connector's configuration options:

| Configuration Option                   | Description                                                             | Default Value |
|---------------------------------------|-------------------------------------------------------------------------|---------------|
| `connect.datalake.kcql`               | Specifies the mapping between Kafka topics and Azure Data Lake using Kafka Connect Query Language. | - |
| `connect.datalake.azure.account.key`   | Azure Account Key for authentication.                                  | [hidden]      |
| `connect.datalake.azure.account.name`  | Azure Account Name for authentication.                                 | ""            |
| `connect.datalake.azure.auth.mode`     | Authentication mode, with options such as 'credentials', 'connectionstring', or 'default'. | Default |
| `connect.datalake.azure.connection.string` | Azure Connection String for authentication.                       | [hidden]      |
| `connect.datalake.compression.codec`   | Compression codec for data formats.                                    | UNCOMPRESSED  |
| `connect.datalake.compression.level`   | Compression level for certain codecs.                                  | -1            |
| `connect.datalake.disable.flush.count` | Option to disable flushing based on record count.                     | false         |
| `connect.datalake.endpoint`            | Azure endpoint for Data Lake.                                           | ""            |
| `connect.datalake.http.connection.timeout` | Connection timeout for HTTP requests.                                | 60000         |
| `connect.datalake.http.socket.timeout`  | Socket timeout for HTTP requests.                                      | 60000         |
| `connect.datalake.local.tmp.directory`  | Local temporary directory for file preparation.                        | ""            |
| `connect.datalake.padding.length`       | Padding length for string values.                                       | -1            |
| `connect.datalake.padding.strategy`     | Padding strategy for partition and offset in output files. Options are 'LeftPad', 'RightPad', or 'NoOp' (no padding). Defaults to 'LeftPad'. | "" |
| `connect.datalake.pool.max.connections` | Maximum connections in the pool. -1: Use default according to the underlying client. | -1 |
| `connect.datalake.error.policy`         | Error handling policy, with options like 'NOOP', 'THROW', and 'RETRY'.   | THROW         |
| `connect.datalake.http.max.retries`     | Number of HTTP request retries for resolvable errors.                  | 5             |
| `connect.datalake.max.retries`         | Maximum number of write retries.                                       | 20            |
| `connect.datalake.retry.interval`      | Time in milliseconds between retry attempts.                           | 60000         |
| `connect.datalake.seek.max.files`       | Maximum index files allowed per topic/partition. Advisable not to raise this, as too many files indicate a problem with file deletion. | 5 |

## Getting Started

To get started with the Kafka sink connector for Azure Data Lake, you'll need to have an Azure account, create an Azure Data Lake Store, and obtain the necessary credentials. Then, configure the connector using the provided options to define the data flow from Kafka to Azure Data Lake.

Make sure to adjust the configuration according to your specific use case and requirements, and ensure that the Azure permissions are set up to allow data access and write operations to your Azure Data Lake storage.

In summary, this connector simplifies the process of streaming data from Kafka to Azure Data Lake, making it an essential tool for building real-time data pipelines and analytics solutions in the Azure ecosystem.

### CODE REVIEW

File: kafka-connect-azure-datalake/src/main/scala/io/lenses/streamreactor/connect/datalake/auth/DatalakeClientCreator.scala [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import io.lenses.streamreactor.connect.datalake.config.AzureConfig
+import io.lenses.streamreactor.connect.datalake.config.ConnectionPoolConfig
+
+import java.time.Duration
+import scala.util.Try
+
+object DatalakeClientCreator extends ClientCreator[AzureConfig, DataLakeServiceClient] {
+
+  def make(config: AzureConfig): Either[Throwable, DataLakeServiceClient] = {
+    require(config != null, "AzureDataLakeConfig cannot be null")
```
Comment (davidsloan) [inline kafka-connect-azure-datalake/src/main/scala/io/lenses/streamreactor/connect/datalake/auth/DatalakeClientCreator.scala:39]: this will cause an exception instead of the either which it should return

**Review by andrewstevenson — APPROVED** [review]
